### ISSUE TITLE
Stablish security guidelines for Component (React) development

### ISSUE THREAD

**Issue by jbalsas** [liferay/liferay-frontend-guidelines#33]
As we move forward, it would be good to stablish some basics Security guidelines and best practices like:
- Sanitizing inputs
- Avoiding certain constructions
- Flagging `dangerouslySetInnerHTML` usages
- ...

These are just examples. The goal of this issue is to discuss and document what constitutes a safe baseline that we can automate and check to avoid security issues in our developments.

**Comment by wincent** [comment]
That's a good list. React does close off quite a few attack vectors by virtue of how it gates DOM access via a layer of indirection, but it's definitely worth continuing to think about [the security aspects](https://medium.com/dailyjs/exploiting-script-injection-flaws-in-reactjs-883fb1fe36c1).

As the name implies `dangerouslySetInnerHTML` isn't safe, so we should probably lint for it. In general I think documenting this kind of thing is good, but automatic enforcement is even better. For that specific instance, the [`react/no-danger` lint rule](https://github.com/yannickcr/eslint-plugin-react/blob/HEAD/docs/rules/no-danger.md) that comes with eslint-plugin-react will be a good start.

For things that can't be automated, it would be nice to present this in the form of a checklist; [here is one example](https://www.sqreen.com/checklists/html-css-security-checklist) that I found via Google. And beyond React, there are quite a few other things that we might want to look at in the frontend space, such as [HTTP headers and such](https://blog.risingstack.com/node-js-security-checklist/).

**Comment by kresimir-coko** [comment]
@wincent 
> For things that can't be automated, it would be nice to present this in the form of a checklist; here is one example that I found via Google.

Oh I really like the checklist, it looks nice and it can serve as inspiration for us to make similar checklists regarding most common processes (like initializing a JS module in DXP) for both us internally and for our customers' devs.

**Comment by kresimir-coko** [comment]
If this doesn't fit under `security` I can open an independant issue for this, otherwise I'll remove this first line.

One of the security problems I encountered is related to `localStorage`.

Here's the [article](https://michalzalecki.com/why-using-localStorage-directly-is-a-bad-idea/) that helped me understand why `localStorage` can be detrimental because it can evaluate as `undefined`.
 
What I did to circumvent the unreliability of `localStorage` was first **testing if it exists** by attempting to set an item to it: `localStorage.setItem()` & `localStorage.removeItem()`. If it successfully sets and then removes the item, that means that `localStorage` is defined, and therefore it can be used in the app. Here's the wrapper I made around `localStorage` to avoid it being `undefined`: https://github.com/liferay/clay/blob/master/clayui.com/src/utils/localStorage.tsx

**Comment by wincent** [comment]
> One of the security problems I encountered is related to `localStorage`.

I don't think this is a "security" problem as such, even though the browser may throw a `SecurityError`; it's really a `YouTriedToUseAnAPIThatCannotBeUsedGivenTheUsersCurrentPrivacySettings`.

Security is about stopping bad actors from doing nasty things with your software, not usually about dealing with API behavior differences that stem from user settings.

**Comment by markocikos** [comment]
`dangerouslySetInnerHTML` is an odd one. There are some legitimate uses for it, like:
1. Localization, where part of a sentence is link. In one way or another, we must continue to support this. I'm not sure if there is a way besides using `dangerouslySetInnerHTML`.
1. Support for long existing (ugly) practice of building HTML in backend. Cost of re-tumbling the whole stack is often too high compared to just adding support in a new frontend component.
1. Support for full customization of parts of UI, for example custom component header. It's a trade-off in sacrificing security for customization capability.

I guess we cannot put this in the linter. At the very least, we should but very careful about it, allowing it in case-to-case basis, when there is no alternative solution.
